drivers/random/random_nrf5.c

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2017 Exati Tecnologia Ltda.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <random.h>
+#include <nrf_rng.h>
+
+static inline u8_t random_nrf5_get_u8(void)
+{
+	u8_t value;
+
+	while (!nrf_rng_event_get(NRF_RNG_EVENT_VALRDY)) {
+	}
+
+	value = nrf_rng_random_value_get();
+	nrf_rng_event_clear(NRF_RNG_EVENT_VALRDY);
+	return value;
+}
+
+static int random_nrf5_get_entropy(struct device *device, u8_t *buf, u16_t len)
+{
+	while (len) {
+		*buf = random_nrf5_get_u8();
+		buf++;
+		len--;
+	}
+
+	return 0;
+}
+
+static int random_nrf5_init(struct device *device)
+{
+	/* Enable or disable bias correction */
+	if (CONFIG_RANDOM_NRF5_BIAS_CORRECTION) {
+		nrf_rng_error_correction_enable();
+	} else {
+		nrf_rng_error_correction_disable();
+	}
+
+	/* Disable the shortcut that stops the task after a byte is generated */
+	nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
+
+	/* Start the nRF5 RNG generator */
+	nrf_rng_task_trigger(NRF_RNG_TASK_START);
```

Comment:
This will make the RNG peripheral running (and consuming power) all the time, regardless of whether random values are needed or not. Consider starting and stopping the peripheral in `random_nrf5_get_entropy`. This will introduce some extra delay before the first random byte is generated, but it does not seem to be a problem.

Reply:
Hi Andrzej,

That's a good point. I was optimizing for performance, but seeing as generating random numbers is fairly rare in a program and nRF52 family is supposed to be low power, I think I was optimizing for the wrong thing. This, and the fact that I didn't want to deal with concurrency... :stuck_out_tongue: 

I implemented your suggestion. Could you take another look?

To make it work, I implemented two mechanisms:
- An atomic user counter, which should solve higher priority threads/interrupts being scheduled, generating random numbers, and then turning off the peripheral, leaving the lower priority thread helpless;
- A timeout, which solves data races regarding turning off the peripheral. After all, we are performing a non-atomic operation after checking for atomic zero. However, I think these should be pretty rare if not impossible, considering RTOSes' scheduling.